File: src/Cpp/examples/example06_mixed_local_remote.cpp
#include <iostream>
#include <string>
#include <utility>

#include "pipeline_services/core/pipeline.hpp"
#include "pipeline_services/examples/text_steps.hpp"
#include "pipeline_services/remote/http_step.hpp"

int main() {
  const std::string fixture_endpoint = "http://127.0.0.1:8765/echo";

  pipeline_services::remote::RemoteSpec<std::string> remote_spec(fixture_endpoint);
  remote_spec.timeoutMillis = 1000;
  remote_spec.retries = 0;

  pipeline_services::core::Pipeline<std::string> pipeline("example06_mixed_local_remote", true);
  pipeline.addAction(pipeline_services::examples::strip);
  pipeline.addAction(pipeline_services::examples::normalize_whitespace);
  pipeline.addAction("remote_echo", pipeline_services::remote::jsonPost<std::string>(std::move(remote_spec)));
  pipeline.addAction(pipeline_services::examples::to_lower);
  pipeline.addAction(pipeline_services::examples::append_marker);

  const std::string output_value = pipeline.run("  Hello   Remote  ");
  std::cout << "output=" << output_value << std::endl;
  return 0;
}

File: src/Rust/examples/example06_mixed_local_remote.rs
use pipeline_services::examples::text_steps::{append_marker, normalize_whitespace, strip, to_lower};
use pipeline_services::remote::http_step::{http_step, RemoteSpec};
use pipeline_services::Pipeline;

fn remote_echo(text_value: String) -> String {
  let fixture_endpoint = "http://127.0.0.1:8765/echo";
  let mut remote_spec = RemoteSpec::new(fixture_endpoint);
  remote_spec.method = "POST".to_string();
  remote_spec.timeout_millis = 1000;
  remote_spec.retries = 0;

  match http_step(&remote_spec, &text_value) {
    Ok(response_body) => response_body,
    Err(message) => panic!("{message}"),
  }
}

fn main() {
  let mut pipeline = Pipeline::new("example06_mixed_local_remote", true);
  pipeline.add_action(strip);
  pipeline.add_action(normalize_whitespace);
  pipeline.add_action_named("remote_echo", remote_echo);
  pipeline.add_action(to_lower);
  pipeline.add_action(append_marker);

  let result = pipeline.run("  Hello   Remote  ".to_string());
  println!("output={}", result.context);
}
